### PR TITLE
feat(cli): add 'drain' CLI command to set /status/ready to 503

### PR DESCRIPTION
**Pull request by tzssangglass**
### Summary

This command updates the `/status/ready` endpoint to return a `503 Service Unavailable` status code. This allows external tools, such as Kubernetes, to detect when Kong is not ready to receive traffic. Based on this response, Kubernetes can gracefully remove Kong from its load balancing pool according to its configured policies, facilitating a smooth shutdown process.
### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix: [FTI-6276](https://konghq.atlassian.net/browse/FTI-6276)

[FTI-6276]: https://konghq.atlassian.net/browse/FTI-6276?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: kong/status/ready.lua [hunk 1]
```diff
@@ -102,6 +103,16 @@ end
 return {
   ["/status/ready"] = {
     GET = function(self, dao, helpers)
+      local unready = kong_shm:get(KONG_STATUS_UNREADY)
+      if unready == nil then
+        kong_shm:set(KONG_STATUS_UNREADY, false)
+      end
+
+      if unready == true then
+        ngx_log(ngx_DEBUG, "not ready for proxying: unready")
```
Comment (windmgc) [inline kong/status/ready.lua:112]:
```suggestion
```

This log line seems ineffective, and could possibly flood the error log if the `/status/ready` is configured as healthcheck endpoint

File: kong/status/ready.lua [hunk 2]
```diff
@@ -112,5 +123,12 @@ return {
         return kong.response.exit(503, { message = err })
       end
     end
+  },
+
+  ["/status/unready"] = {
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply kong/status/ready.lua:128]: We do not have a solid need to set Kong in a ready state manually; the `/status/ready` endpoint(GET) is enough.

The `/status/unready` endpoint is used for some admin systems (like k8s) to make `/status/ready` return 503.

File: kong/status/ready.lua [hunk 3]
```diff
@@ -23,6 +23,7 @@ local PLUGINS_REBUILD_COUNTER_KEY = constants.PLUGINS_REBUILD_COUNTER_KEY
 local ROUTERS_REBUILD_COUNTER_KEY = constants.ROUTERS_REBUILD_COUNTER_KEY
 local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
 
+local KONG_STATUS_UNREADY = "kong:status:unready"
```
[the remark this answers is not in the document]
Comment (tzssangglass) [reply kong/status/ready.lua]: done

File: kong/status/ready.lua (hunk 1 above)
Comment (tzssangglass) [reply kong/status/ready.lua]: done

File: kong/status/ready.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (windmgc) [reply kong/status/ready.lua:127]: Discussed offline, it'd be better to put it into a POST `/status/ready` method with status argument, and called by a CLI command named `kong drain`

[the remark this answers is not in the document]
Comment (tzssangglass) [reply kong/status/ready.lua]: done

File: kong/status/ready.lua [hunk 4]
```diff
@@ -111,6 +121,15 @@ return {
...
       end
+    end,
+
+    POST = function(self, dao, helpers)
+      if self.params and self.params.status == "draining" then
+        kong_shm:set(KONG_STATUS_READY, false)
+        return kong.response.exit(204)
+      end
+
+      return kong.response.exit(501)
```
Comment (windmgc) [inline kong/status/ready.lua:132]:
```suggestion
      return kong.response.exit(400)
```
Feels like a 400 response is more appropriate

**Review by windmgc — APPROVED** [review]
Approve with one minor suggestion

**Comment by tzssangglass** [comment]
hi @mheap , PTAL this PR, it add a new command `drain` in Kong.

File: spec/02-integration/02-cmd/17-drain_spec.lua [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  if not res then
+    return nil, err
+  end
+
+  return res.status
+end
+
+for _, strategy in helpers.each_strategy() do
+  if strategy ~= "off" then
+    -- skip the "off" strategy, as dbless has its own test suite
```
Comment (mheap) [inline spec/02-integration/02-cmd/17-drain_spec.lua:36]: Where is the dbless test suite? If this is used with k8s, it needs to support dbless

**Review by gszr — COMMENTED** [review]
@tzssangglass Is there an internal spike doc with the alternatives we looked into for solving this issue? I am curious if adding yet another command isn't overcomplicating things -- maybe it's the best option, but I'm curious if this is the best interface we can provide among the options we researched. Adding new commands to the CLI increases complexity in a non-negligible way.

File: spec/02-integration/02-cmd/17-drain_spec.lua (hunk 5 above)
Comment (tzssangglass) [reply spec/02-integration/02-cmd/17-drain_spec.lua:51]: add DB-less test case

**Comment by tzssangglass** [comment]
> Is there an internal spike doc with the alternatives we looked into for solving this issue?

We do not have an internal spike doc recording this or an alternative solution; 

This was derived from discussions with members of the k8s team.

**Review by mheap — APPROVED** [review]

**Pull request merged**